File: 프로그래머스/weeklyChallenge/006.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> W;
vector<vector<int>> order;

bool cmp(vector<int> a, vector<int> b){
    float A = 0;
    float B = 0;
    if(a[2] != 0){
        A = (float)a[3] / (float)a[2];
    }
    if(b[2] != 0){
        B = (float)b[3] / (float)b[2];
    }
    if(A == B){
        if(a[1] == b[1]){
            if(W[a[0]] == W[b[0]]){
                return a[0] < b[0];
            }
            return W[a[0]] > W[b[0]];
        }
        return a[1] > b[1];
    }
    return A > B;
}

vector<int> solution(vector<int> weights, vector<string> head2head) {
    vector<int> answer;
    W = weights;
    for(int i=0;i<head2head.size();i++){
        int win = 0;
        int game = 0;
        int wintotal = 0;
        vector<int> tmp;
        for(int j=0;j<head2head[i].size();j++){
            if(head2head[i][j] == 'W' && weights[i] < weights[j]){
                win ++;
            }
            if(head2head[i][j] != 'N'){
                game ++;
            }
            if(head2head[i][j] == 'W'){
                wintotal ++;
            }
        }
        tmp.push_back(i);
        tmp.push_back(win);
        tmp.push_back(game);
        tmp.push_back(wintotal);
        order.push_back(tmp);
    }
    sort(order.begin(),order.end(),cmp);
    for(int i=0;i<order.size();i++){
        //cout << order[i][0] << " " << order[i][1] << " " << order[i][2] << " " << order[i][3] << "\n";
        answer.push_back(order[i][0] + 1);
    }
    return answer;
}

File: 프로그래머스/weeklyChallenge/006.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int[] solution(int[] weights, string[] head2head) {
        int[] answer = new int[weights.Length];
        List<List<int>> Player = new List<List<int>>();
        for(int i=0;i<weights.Length;i++){
            Player.Add(new List<int>());
        }
        for(int i=0;i<head2head.Length;i++){
            int playNum = 0;
            int winNum = 0;
            int heavyWinNum = 0;
            for(int j=0;j<head2head[i].Length;j++){
                if(head2head[i][j] == 'W' || head2head[i][j] == 'L'){
                    playNum ++;
                    if(head2head[i][j] == 'W'){
                        winNum ++;
                        if(weights[i] < weights[j]){
                            heavyWinNum ++;
                        }
                    }
                }
                //Console.Write(head2head[i][j]+" ");
            }
            Player[i].Add(i);
            Player[i].Add(playNum);
            Player[i].Add(winNum);
            Player[i].Add(heavyWinNum);
            Player[i].Add(weights[i]);
            //Console.WriteLine();
        }
        /*
        for(int i=0;i<Player.Count;i++){
            for(int j=0;j<Player[i].Count;j++){
                Console.Write(Player[i][j]+" ");
            }
            Console.WriteLine();
        }
        */
        Console.WriteLine();
        Player.Sort((A,B) => {
            float aWinRate = 0;
            float bWinRate = 0;
            if(A[1] != 0){
                aWinRate = (float)A[2] / (float)A[1];
            }
            if(B[1] != 0){
                bWinRate = (float)B[2] / (float)B[1];
            }
            
            if(aWinRate > bWinRate){
                return -1;
            }else if(aWinRate < bWinRate){
                return 1;
            }else{  // 승률이 같을때,
                if(A[3] > B[3]){
                    return -1;
                }else if(A[3] < B[3]){
                    return 1;
                }else{  // 자신보다 몸무게가 무거운 복서를 이긴 횟수가 같을때
                    if(A[4] > B[4]){
                        return -1;
                    }else if(A[4] < B[4]){
                        return 1;
                    }else{  // 몸무게가 같을 때
                        if(A[0] < B[0]){
                            return -1;
                        }else{
                            return 1;
                        }
                    }
                }
            }
        });
        /*
        for(int i=0;i<Player.Count;i++){
            for(int j=0;j<Player[i].Count;j++){
                Console.Write(Player[i][j]+" ");
            }
            Console.WriteLine();
        }
        */
        
        for(int i=0;i<Player.Count;i++){
            answer[i] = (Player[i][0]+1);
        }
        return answer;
    }
}